modules/apps/bean-portlet/bean-portlet-spring-extension/src/main/java/com/liferay/bean/portlet/spring/extension/internal/SpringBeanPortletExtension.java

```diff
@@ -0,0 +1,563 @@
+
+								if (Validator.isNull(beanMethodResourceID) ||
+									beanMethodResourceID.equals(resourceID)) {
+
+									ResourceResponse resourceResponse =
+										(ResourceResponse)portletResponse;
+
+									String contentType =
+										beanPortletMethodType.getContentType(
+											method);
+
+									if (Validator.isNotNull(contentType) &&
+										!Objects.equals(contentType, "*/*")) {
+
+										resourceResponse.setContentType(
+											contentType);
+									}
+
+									String characterEncoding =
+										beanPortletMethodType.
+											getCharacterEncoding(method);
+
+									if (Validator.isNotNull(
+											characterEncoding)) {
+
+										resourceResponse.setCharacterEncoding(
+											characterEncoding);
+									}
+
+									if (method.getParameterCount() == 0) {
+										String markup =
+											(String)beanMethod.invoke();
+
+										if (Validator.isNotNull(markup)) {
+											PrintWriter printWriter =
+												resourceResponse.getWriter();
+
+											printWriter.write(markup);
+										}
+									}
+									else {
+										beanMethod.invoke(
+											resourceRequest, resourceResponse);
+									}
+
+									include = beanPortletMethodType.getInclude(
+										method);
+								}
+							}
+							else {
+								beanMethod.invoke(
+									portletRequest, portletResponse);
+							}
+
+							PortletMode beanMethodPortletMode =
+								beanMethod.getPortletMode();
+
+							if (Validator.isNotNull(include) &&
+								((beanMethodPortletMode == null) ||
+								 beanMethodPortletMode.equals(
+									 portletRequest.getPortletMode()))) {
+
+								PortletContext portletContext =
+									portletConfig.getPortletContext();
+
+								PortletRequestDispatcher requestDispatcher =
+									portletContext.getRequestDispatcher(
+										include);
+
+								if (requestDispatcher == null) {
+									_log.error(
+										"Unable to acquire dispatcher to " +
+											"include=" + include);
+								}
+								else {
+									requestDispatcher.include(
+										portletRequest, portletResponse);
+								}
+							}
+						}
+						catch (InvocationTargetException ite) {
+							Throwable cause = ite.getCause();
+
+							if (cause instanceof PortletException) {
+								throw (PortletException)cause;
+							}
+
+							throw new PortletException(cause);
+						}
+						catch (PortletException pe) {
+							throw pe;
+						}
+						catch (Exception e) {
+							throw new PortletException(e);
+						}
+					}
+
+					// MVC
+
+					if (portletResponse instanceof RenderResponse ||
+						portletResponse instanceof ResourceResponse) {
+
+						ViewRenderer viewRenderer = _applicationContext.getBean(
+							"viewRenderer", ViewRenderer.class);
+
+						viewRenderer.render(
+							portletRequest, (MimeResponse)portletResponse,
+							portletConfig);
+					}
+				}
+
+			};
+
+		_beanPortletRegistrar = (BeanPortletRegistrar)bundleContext.getService(
```

Comment:
@ngriffin7a is this the reason why all modules are static? You are assuming the BeanPortletRegistrar service is statically available.

Reply:
@shuyangzhou When you write "static" do you mean vs dynamic? If so, is there a way to do dynamic other than injection via `@Reference`? Maybe you could provide me with a link to a line of code somewhere that shows the dynamic syntax you are asking about.

Reply:
I meant static bundles, all your bundles are static, are you doing this because you need to assume BeanPortletRegistrarImpl is registered at very early stage?

Reply:
@shuyangzhou Oh, you're referring to them living in $LIFERAY_HOME/osgi/static? I did not do that intentionally. I must have simply copied patterns that were already there.

BTW, I would like to mention that com.liferay.bean.portlet.spring.extension.jar is currently designed to live in WEB-INF/lib -- it is not designed to be a static or dynamic OSGi module.

Reply:
Ok, I will try to make all module non-static, as that will cause them to be initialized earlier than necessary, which will slow down startup.

Reply:
Good catch. Sounds good, thank you.